File: impl/go/sim/parallel.go
// Multi-threaded tick (SPEC-1 section 5.6, benchmark class P).
//
// Both reduction strategies, the same phase order and the same deposit order
// as the C reference, so `binned` is bit-identical to the single-threaded run
// here too. The agent rule and the stencil are not reimplemented: this file
// calls Sim.AgentRange and Sim.DiffuseRows, the same code the serial path uses.
//
// # The barrier
//
// Go's standard library has no barrier. sync.WaitGroup is the obvious
// substitute -- Add(n), each worker Done(), everyone Wait() -- but a WaitGroup
// is single-use, so a six-phase tick would allocate and re-Add six of them per
// tick, and the workers would need a second signal to know the next phase had
// started. The sense-reversing barrier below is one mutex and one condition
// variable, reused for the whole run.
//
// Goroutines are pinned to nothing: Go has no affinity API, and GOMAXPROCS is
// set to the thread count so the scheduler has exactly as many Ps as workers.
package sim

import (
	"os"
	"sync"
	"time"
)

type barrier struct {
	n     int
	mu    sync.Mutex
	cv    *sync.Cond
	count int
	sense bool
}

func newBarrier(n int) *barrier {
	b := &barrier{n: n}
	b.cv = sync.NewCond(&b.mu)
	return b
}

func (b *barrier) wait() {
	b.mu.Lock()
	my := !b.sense
	b.count++
	if b.count == b.n {
		b.count = 0
		b.sense = my
		b.cv.Broadcast()
	} else {
		for b.sense != my {
			b.cv.Wait()
		}
	}
	b.mu.Unlock()
}

// split is a contiguous split of [0,n) into `parts`; part i is [lo,hi).
// Identical to the C reference's split, deliberately: the partition is what
// makes `binned` reproduce the serial deposit order.
func split(n, parts, i int) (int, int) {
	base, rem := n/parts, n%parts
	lo := i*base + min(i, rem)
	hi := lo + base
	if i < rem {
		hi++
	}
	return lo, hi
}

type pool struct {
	s        *Sim
	t        int
	cells    int
	agents   int
	height   int
	log2w    uint32
	deposit  float32
	reduce   Reduce
	adaptive bool
	bar      *barrier

	aidx    []int32
	sorted  []int32
	counts  []int32
	offsets []int32
	ybucket []int32
	rowcnt  []int32
	rowsum  []int32
	priv    []float32
}

func (p *pool) bucketOf(idx int32) int { return int(p.ybucket[int(idx)>>p.log2w]) }

// ---- Reduce::Private -------------------------------------------------------

func (p *pool) agentsPrivate(tid int) {
	lo, hi := split(p.agents, p.t, tid)
	p.s.AgentRange(lo, hi, p.aidx)
	base := tid * p.cells
	for i := lo; i < hi; i++ {
		p.priv[base+int(p.aidx[i])] += p.deposit
	}
}

// mergePrivate uses a fixed worker order, so the result is reproducible for
// this worker count. It is NOT in general the same grouping as the serial
// chain -- SPEC-1 section 5.6.
func (p *pool) mergePrivate(tid int) {
	lo, hi := split(p.cells, p.t, tid)
	g := p.s.Grid
	for i := lo; i < hi; i++ {
		acc := p.priv[i]
		p.priv[i] = 0
		for k := 1; k < p.t; k++ {
			acc += p.priv[k*p.cells+i]
			p.priv[k*p.cells+i] = 0
		}
		g[i] += acc
	}
}

// ---- Reduce::Binned --------------------------------------------------------

func (p *pool) agentsBinned(tid int) {
	lo, hi := split(p.agents, p.t, tid)
	cnt := p.counts[tid*p.t : (tid+1)*p.t]
	for i := range cnt {
		cnt[i] = 0
	}
	var rc []int32
	if p.adaptive {
		rc = p.rowcnt[tid*p.height : (tid+1)*p.height]
		for i := range rc {
			rc[i] = 0
		}
	}

	p.s.AgentRange(lo, hi, p.aidx)

	for i := lo; i < hi; i++ {
		idx := p.aidx[i]
		cnt[p.bucketOf(idx)]++
		if p.adaptive {
			rc[int(idx)>>p.log2w]++
		}
	}
}

// prefixBinned is the prefix sum over (bucket, worker) in that order, by
// worker 0 alone. Because each worker owns a contiguous ascending agent range,
// walking workers in order inside a bucket lays the agents down in ascending
// global index -- which is what makes the deposit chain identical to the
// serial one.
func (p *pool) prefixBinned() {
	running := int32(0)
	for b := 0; b < p.t; b++ {
		for w := 0; w < p.t; w++ {
			p.offsets[w*p.t+b] = running
			running += p.counts[w*p.t+b]
		}
	}
}

func (p *pool) scatterBinned(tid int) {
	lo, hi := split(p.agents, p.t, tid)
	off := p.offsets[tid*p.t : (tid+1)*p.t]
	for i := lo; i < hi; i++ {
		b := p.bucketOf(p.aidx[i])
		p.sorted[off[b]] = int32(i)
		off[b]++
	}
}

// depositBinned applies exactly the deposits landing in this worker's row
// block, in ascending agent index. Cells in other blocks are never touched.
func (p *pool) depositBinned(tid int) {
	begin := 0
	for b := 0; b < tid; b++ {
		for w := 0; w < p.t; w++ {
			begin += int(p.counts[w*p.t+b])
		}
	}
	end := begin
	for w := 0; w < p.t; w++ {
		end += int(p.counts[w*p.t+tid])
	}
	dep := p.s.Dep
	for j := begin; j < end; j++ {
		dep[p.aidx[p.sorted[j]]] += p.deposit
	}
}

func (p *pool) mergeBinned(tid int) {
	ylo, yhi := split(p.height, p.t, tid)
	p.s.MergeRows(ylo, yhi)
	if !p.adaptive {
		return
	}
	for y := ylo; y < yhi; y++ {
		var sum int32
		for w := 0; w < p.t; w++ {
			sum += p.rowcnt[w*p.height+y]
		}
		p.rowsum[y] = sum
	}
}

// rebalance recomputes row boundaries so every worker gets a similar number of
// deposits. It cannot change the result: the partition decides which worker
// applies a deposit, never the order deposits hit a cell.
func (p *pool) rebalance() {
	var total int64
	for _, v := range p.rowsum {
		total += int64(v)
	}
	if total == 0 {
		return
	}
	b, acc := int32(0), int64(0)
	for y := 0; y < p.height; y++ {
		p.ybucket[y] = b
		acc += int64(p.rowsum[y])
		for int(b)+1 < p.t && acc*int64(p.t) >= total*int64(b+1) &&
			p.height-y-1 >= p.t-int(b)-1 {
			b++
		}
	}
}

// runTick is one tick, run by every worker. Six barriers for `binned`,
// matching the C reference's five phase barriers plus its handshake. The
// rebalance overlaps the diffusion pass, which does not read ybucket.
func (p *pool) runTick(tid int) {
	if p.reduce == Binned {
		p.agentsBinned(tid)
		p.bar.wait()
		if tid == 0 {
			p.prefixBinned()
		}
		p.bar.wait()
		p.scatterBinned(tid)
		p.bar.wait()
		p.depositBinned(tid)
		p.bar.wait()
		p.mergeBinned(tid)
		p.bar.wait()
		if tid == 0 && p.adaptive {
			p.rebalance()
		}
	} else {
		p.agentsPrivate(tid)
		p.bar.wait()
		p.mergePrivate(tid)
		p.bar.wait()
	}

	ylo, yhi := split(p.height, p.t, tid)
	p.s.DiffuseRows(ylo, yhi)
	p.bar.wait()

	// The two slice headers are shared, so exactly one worker swaps them, and
	// it does so between two barriers -- everyone else is parked.
	if tid == 0 {
		p.s.SwapBuffers()
	}
	p.bar.wait()
}

// Run is the result of a class-P run.
type Run struct {
	MsTotal float64
	TickMs  []float64
}

// RunParallel runs warmup+ticks ticks across cfg.Threads goroutines.
func RunParallel(s *Sim, warmup, ticks uint32) (*Run, error) {
	c := s.Cfg
	if c.Update != Deferred {
		return nil, simError(
			"--threads > 1 requires --update deferred.\n" +
				"       SPEC-1 'serial' makes an agent's deposit visible to the\n" +
				"       next agent in the same tick, which is a sequential\n" +
				"       dependency; see SPEC-1 section 5.5.")
	}
	t := int(c.Threads)
	cells := int(c.Width) * int(c.Height)
	h := int(c.Height)
	binned := c.Reduce == Binned

	p := &pool{
		s: s, t: t, cells: cells, agents: int(c.Agents), height: h,
		log2w: s.log2w, deposit: c.Deposit, reduce: c.Reduce,
		adaptive: binned && os.Getenv("SLIMEBENCH_NO_REBALANCE") == "",
		bar:      newBarrier(t),
		aidx:     make([]int32, c.Agents),
	}
	if binned {
		p.sorted = make([]int32, c.Agents)
		p.counts = make([]int32, t*t)
		p.offsets = make([]int32, t*t)
		p.ybucket = make([]int32, h)
		// Row -> owning worker, same split as the diffusion pass so a worker's
		// deposits land in rows it already touches.
		for b := 0; b < t; b++ {
			lo, hi := split(h, t, b)
			for y := lo; y < hi; y++ {
				p.ybucket[y] = int32(b)
			}
		}
		if p.adaptive {
			p.rowcnt = make([]int32, t*h)
			p.rowsum = make([]int32, h)
		}
	} else {
		p.priv = make([]float32, t*cells)
	}

	total := int(warmup + ticks)
	var wg sync.WaitGroup
	for tid := 1; tid < t; tid++ {
		wg.Add(1)
		go func(tid int) {
			defer wg.Done()
			for i := 0; i < total; i++ {
				p.runTick(tid)
			}
		}(tid)
	}

	// Worker 0 is a worker like the others; it just also holds the clock.
	// runTick ends with a barrier, so reading it here measures the whole pool
	// rather than worker 0's own progress.
	for i := uint32(0); i < warmup; i++ {
		p.runTick(0)
	}
	tickMs := make([]float64, 0, ticks)
	start := time.Now()
	for i := uint32(0); i < ticks; i++ {
		a := time.Now()
		p.runTick(0)
		tickMs = append(tickMs, float64(time.Since(a).Nanoseconds())/1e6)
	}
	ms := float64(time.Since(start).Nanoseconds()) / 1e6
	wg.Wait()

	// The phases interleave, so an agent/diffuse split would be meaningless.
	s.NsAgents = int64(ms * 1e6)
	s.NsDiff = 0
	return &Run{MsTotal: ms, TickMs: tickMs}, nil
}

File: impl/swift/Sources/SlimebenchCore/Parallel.swift
// Multi-threaded tick (SPEC-1 section 5.6, benchmark class P).
//
// Both reduction strategies, the same phase order and the same deposit order
// as the C reference, so `binned` is bit-identical to the single-threaded run
// here too. The agent rule and the stencil are not reimplemented: this file
// calls Sim.agentRange and Sim.diffuseRows, the same code the serial path uses.
//
// # Threads, not Dispatch
//
// The obvious Swift answer is `DispatchQueue.concurrentPerform`, and it is the
// wrong one here: it is a fork-join over one phase, so a six-phase tick means
// six fork-joins per tick, each re-entering the thread pool. What this needs is
// six *barriers* between workers that stay alive for the whole run, which is
// `Foundation.Thread` plus a condition variable.
//
// Swift 6's strict concurrency does not model this -- the workers share one
// mutable grid and coordinate by barrier, which is exactly what the actor model
// is designed to prevent. `@unchecked Sendable` is the honest annotation: the
// safety argument is the barrier, and it is written out below rather than
// checked by the compiler.

import Foundation

final class Barrier {
    private let n: Int
    private let cond = NSCondition()
    private var count = 0
    private var sense = false

    init(_ n: Int) { self.n = n }

    func wait() {
        cond.lock()
        let my = !sense
        count += 1
        if count == n {
            count = 0
            sense = my
            cond.broadcast()
        } else {
            while sense != my { cond.wait() }
        }
        cond.unlock()
    }
}

/// Contiguous split of [0, n) into `parts`; part `i` is [lo, hi). Identical to
/// the C reference's `split`: the partition is what makes `binned` reproduce
/// the serial deposit order.
@inline(__always)
func splitRange(_ n: Int, _ parts: Int, _ i: Int) -> (Int, Int) {
    let base = n / parts, rem = n % parts
    let lo = i * base + Swift.min(i, rem)
    return (lo, lo + base + (i < rem ? 1 : 0))
}

/// Everything the workers share. Unchecked because the invariant that makes it
/// safe -- disjoint index ranges per phase, phases separated by barriers -- is
/// not one the compiler can see.
final class Pool: @unchecked Sendable {
    let sim: Sim
    let t: Int
    let cells: Int
    let agents: Int
    let height: Int
    let log2w: UInt32
    let deposit: Float
    let reduce: Reduce
    let adaptive: Bool
    let bar: Barrier

    var aidx: UnsafeMutablePointer<Int32>
    var sorted: UnsafeMutablePointer<Int32>
    var counts: UnsafeMutablePointer<Int32>
    var offsets: UnsafeMutablePointer<Int32>
    var ybucket: UnsafeMutablePointer<Int32>
    var rowcnt: UnsafeMutablePointer<Int32>
    var rowsum: UnsafeMutablePointer<Int32>
    var priv: UnsafeMutablePointer<Float>

    init(_ sim: Sim, adaptive: Bool) {
        self.sim = sim
        let c = sim.cfg
        t = Int(c.threads)
        cells = Int(c.width) * Int(c.height)
        agents = Int(c.agents)
        height = Int(c.height)
        log2w = sim.log2w
        deposit = c.deposit
        reduce = c.reduce
        self.adaptive = adaptive && c.reduce == .binned
        bar = Barrier(t)

        let binned = c.reduce == .binned
        func alloc32(_ n: Int) -> UnsafeMutablePointer<Int32> {
            let p = UnsafeMutablePointer<Int32>.allocate(capacity: Swift.max(1, n))
            p.initialize(repeating: 0, count: Swift.max(1, n))
            return p
        }
        aidx = alloc32(agents)
        sorted = alloc32(binned ? agents : 1)
        counts = alloc32(binned ? t * t : 1)
        offsets = alloc32(binned ? t * t : 1)
        ybucket = alloc32(binned ? height : 1)
        rowcnt = alloc32(binned && self.adaptive ? t * height : 1)
        rowsum = alloc32(binned && self.adaptive ? height : 1)
        let pn = binned ? 1 : t * cells
        priv = UnsafeMutablePointer<Float>.allocate(capacity: pn)
        priv.initialize(repeating: 0, count: pn)

        if binned {
            // Row -> owning worker, same split as the diffusion pass so a
            // worker's deposits land in rows it already touches.
            for b in 0..<t {
                let (lo, hi) = splitRange(height, t, b)
                for y in lo..<hi { ybucket[y] = Int32(b) }
            }
        }
    }

    deinit {
        aidx.deallocate(); sorted.deallocate(); counts.deallocate()
        offsets.deallocate(); ybucket.deallocate(); rowcnt.deallocate()
        rowsum.deallocate(); priv.deallocate()
    }

    @inline(__always)
    func bucketOf(_ idx: Int32) -> Int { Int(ybucket[Int(idx) >> log2w]) }

    // ---- private ----------------------------------------------------------

    func agentsPrivate(_ tid: Int) {
        let (lo, hi) = splitRange(agents, t, tid)
        sim.agentRange(lo, hi, aidx)
        let base = tid * cells
        for i in lo..<hi { priv[base + Int(aidx[i])] += deposit }
    }

    /// Fixed worker order, so the result is reproducible for this worker
    /// count. It is NOT in general the same grouping as the serial chain.
    func mergePrivate(_ tid: Int) {
        let (lo, hi) = splitRange(cells, t, tid)
        sim.grid.withUnsafeMutableBufferPointer { gp in
            let g = gp.baseAddress!
            for i in lo..<hi {
                var acc = priv[i]
                priv[i] = 0
                for k in 1..<t {
                    acc += priv[k * cells + i]
                    priv[k * cells + i] = 0
                }
                g[i] += acc
            }
        }
    }

    // ---- binned -----------------------------------------------------------

    func agentsBinned(_ tid: Int) {
        let (lo, hi) = splitRange(agents, t, tid)
        for b in 0..<t { counts[tid * t + b] = 0 }
        if adaptive { for y in 0..<height { rowcnt[tid * height + y] = 0 } }

        sim.agentRange(lo, hi, aidx)

        for i in lo..<hi {
            let idx = aidx[i]
            counts[tid * t + bucketOf(idx)] += 1
            if adaptive { rowcnt[tid * height + (Int(idx) >> log2w)] += 1 }
        }
    }

    /// Prefix sum over (bucket, worker) in that order, by worker 0 alone.
    /// Because each worker owns a contiguous ascending agent range, walking
    /// workers in order inside a bucket lays the agents down in ascending
    /// global index -- which is what makes the deposit chain identical to the
    /// serial one.
    func prefixBinned() {
        var running: Int32 = 0
        for b in 0..<t {
            for w in 0..<t {
                offsets[w * t + b] = running
                running += counts[w * t + b]
            }
        }
    }

    func scatterBinned(_ tid: Int) {
        let (lo, hi) = splitRange(agents, t, tid)
        for i in lo..<hi {
            let b = tid * t + bucketOf(aidx[i])
            sorted[Int(offsets[b])] = Int32(i)
            offsets[b] += 1
        }
    }

    /// Applies exactly the deposits landing in this worker's row block, in
    /// ascending agent index. Cells in other blocks are never touched.
    func depositBinned(_ tid: Int) {
        var begin = 0
        for b in 0..<tid { for w in 0..<t { begin += Int(counts[w * t + b]) } }
        var end = begin
        for w in 0..<t { end += Int(counts[w * t + tid]) }
        guard end > begin else { return }
        sim.dep.withUnsafeMutableBufferPointer { dp in
            let d = dp.baseAddress!
            for j in begin..<end { d[Int(aidx[Int(sorted[j])])] += deposit }
        }
    }

    func mergeBinned(_ tid: Int) {
        let (ylo, yhi) = splitRange(height, t, tid)
        sim.mergeRows(ylo, yhi)
        guard adaptive else { return }
        for y in ylo..<yhi {
            var sum: Int32 = 0
            for w in 0..<t { sum += rowcnt[w * height + y] }
            rowsum[y] = sum
        }
    }

    /// Recompute row boundaries so every worker gets a similar number of
    /// deposits. Cannot change the result: the partition decides *which*
    /// worker applies a deposit, never the order deposits hit a cell.
    func rebalance() {
        var total: Int64 = 0
        for y in 0..<height { total += Int64(rowsum[y]) }
        guard total > 0 else { return }
        var b: Int32 = 0
        var acc: Int64 = 0
        for y in 0..<height {
            ybucket[y] = b
            acc += Int64(rowsum[y])
            while Int(b) + 1 < t && acc * Int64(t) >= total * Int64(b + 1)
                && height - y - 1 >= t - Int(b) - 1 {
                b += 1
            }
        }
    }

    /// One tick, run by every worker. Six barriers for `binned`, matching the
    /// C reference's five phase barriers plus its handshake. The rebalance
    /// overlaps the diffusion pass, which does not read `ybucket`.
    func runTick(_ tid: Int) {
        if reduce == .binned {
            agentsBinned(tid); bar.wait()
            if tid == 0 { prefixBinned() }; bar.wait()
            scatterBinned(tid); bar.wait()
            depositBinned(tid); bar.wait()
            mergeBinned(tid); bar.wait()
            if tid == 0 && adaptive { rebalance() }
        } else {
            agentsPrivate(tid); bar.wait()
            mergePrivate(tid); bar.wait()
        }

        let (ylo, yhi) = splitRange(height, t, tid)
        sim.diffuseRows(ylo, yhi)
        bar.wait()

        // The two arrays are shared, so exactly one worker swaps them, and it
        // does so between two barriers -- everyone else is parked.
        if tid == 0 { sim.swapBuffers() }
        bar.wait()
    }
}

public struct ParallelRun {
    public let msTotal: Double
    public let tickMs: [Double]
}

public func runParallel(_ sim: Sim, warmup: UInt32, ticks: UInt32) throws -> ParallelRun {
    guard sim.cfg.update == .deferred else { throw SimError.serialNotParallel }

    let adaptive = ProcessInfo.processInfo.environment["SLIMEBENCH_NO_REBALANCE"] == nil
    let pool = Pool(sim, adaptive: adaptive)
    let total = Int(warmup + ticks)
    let t = Int(sim.cfg.threads)

    var threads: [Thread] = []
    for tid in 1..<t {
        let th = Thread {
            for _ in 0..<total { pool.runTick(tid) }
        }
        // The default 512 KiB is plenty; naming them makes `top -H` readable.
        th.name = "slimebench-\(tid)"
        th.start()
        threads.append(th)
    }

    // Worker 0 is a worker like the others; it just also holds the clock.
    // runTick ends with a barrier, so reading it here measures the whole pool.
    for _ in 0..<warmup { pool.runTick(0) }
    sim.nsAgents = 0
    sim.nsDiffuse = 0

    var tickMs: [Double] = []
    tickMs.reserveCapacity(Int(ticks))
    let start = DispatchTime.now().uptimeNanoseconds
    for _ in 0..<ticks {
        let a = DispatchTime.now().uptimeNanoseconds
        pool.runTick(0)
        tickMs.append(Double(DispatchTime.now().uptimeNanoseconds - a) / 1e6)
    }
    let ms = Double(DispatchTime.now().uptimeNanoseconds - start) / 1e6

    // Thread has no join; the last barrier of the last tick already released
    // everyone, so spin briefly until they have actually exited.
    for th in threads { while !th.isFinished { usleep(200) } }

    // The phases interleave, so an agent/diffuse split would be meaningless.
    sim.nsAgents = Int(ms * 1e6)
    sim.nsDiffuse = 0
    return ParallelRun(msTotal: ms, tickMs: tickMs)
}
